backends-velox/src/main/scala/org/apache/spark/sql/expression/UDFResolver.scala

```diff
@@ -80,17 +94,132 @@ object UDFResolver extends Logging {
     logInfo(s"Registered UDF: $name -> $t")
   }
 
+  def parseName(name: String): (String, String) = {
+    val index = name.lastIndexOf("#")
+    if (index == -1) {
+      (name, Paths.get(name).getFileName.toString)
+    } else {
+      (name.substring(0, index), name.substring(index + 1))
+    }
+  }
+
+  def getFilesWithExtension(directory: Path, extension: String): Seq[Path] = {
+    Files
+      .walk(directory)
+      .filter(p => Files.isRegularFile(p) && p.toString.endsWith(extension))
+      .iterator()
+      .asScala
+      .toSeq
+  }
+
+  def unpackAndGetLibraries(source: File, destDir: File): Seq[String] = {
+    val sourceName = source.getName
+    val dest = new File(destDir, sourceName)
+    logInfo(
+      s"Unpacking an archive $sourceName from ${source.getAbsolutePath} to ${dest.getAbsolutePath}")
+    Utils.deleteRecursively(dest)
+    Utils.unpack(source, dest)
+    getFilesWithExtension(dest.toPath, LIB_EXTENSION).map(_.toString)
+  }
+
+  def getAllLibraries(files: String): Seq[String] = {
+    getAllLibraries(files.split(","))
+  }
+
+  def getAllLibraries(files: Seq[String]): Seq[String] = {
+    files
+      .map {
+        f =>
+          val file = new File(f)
+          if (file.isAbsolute) {
+            file
+          } else {
+            new File(SparkFiles.get(f))
+          }
+      }
+      .flatMap {
+        f =>
+          if (f.isDirectory) {
+            getFilesWithExtension(f.toPath, LIB_EXTENSION).map(_.toString)
+          } else {
+            Seq(f.toString)
+          }
+      }
+  }
+
   def loadAndGetFunctionDescriptions: Seq[(FunctionIdentifier, ExpressionInfo, FunctionBuilder)] = {
-    SparkContext.getActive.get.conf
-      .getOption(BackendSettings.GLUTEN_VELOX_UDF_LIB_PATHS)
-      .foreach(new UdfJniWrapper().nativeLoadUdfLibraries(_))
-
-    UDFMap.map {
-      case (name, t) =>
-        (
-          new FunctionIdentifier(name),
-          new ExpressionInfo(classOf[UDFExpression].getName, name),
-          (e: Seq[Expression]) => UDFExpression(name, t.dataType, t.nullable, e))
-    }.toSeq
+    val sparkContext = SparkContext.getActive.get
+    val sparkConf = sparkContext.conf
+    sparkConf.getOption(BackendSettings.GLUTEN_VELOX_UDF_LIBS).map(_.split(",")) match {
+      case None =>
+        Seq.empty
+      case Some(udfFiles) =>
+        val master = sparkConf.getOption("spark.master")
+        val isYarn = master.isDefined && master.get.equals("yarn")
+        localLibraryPaths = if (isYarn) {
+          // For Yarn-client or Yarn-cluster mode,
+          // driver cannot get uploaded files via SparkFiles.get.
+          // So we need to copy and unpack files for driver first.
+          val hadoopConf = new YarnConfiguration(SparkHadoopUtil.newConfiguration(sparkConf))
```

Comment:
AFAIK, for yarn-cluster mode, the archives can be unpacked on driver side, so we only need pass the unpaced directory, such as `./libs`, to `udfLibraryPaths`. As for yarn-client mode, why not let user to unpack the archives and pass the directories to `udfLibraryPaths`. 

BTW, if the `udfLibraryPaths` is different between driver side and executor side, we only need to add a `driver.udfLibraryPaths`, which is defauled with `udfLibraryPaths`, to solve this problem. Just like the `spark.pyspark.driver.python` and `spark.pyspark.python`.

Reply:
Sounds great. Will do.